source/MRViewer/MRSceneCache.h

```diff
@@ -14,24 +16,77 @@ class MRVIEWER_CLASS SceneCache
     // call it in the beginning each frame
     MRVIEWER_API static void invalidateAll();
 
-    // get all selectable object in tree
-    // same as getAllObjectsInTree<Object>( &SceneRoot::get(), ObjectSelectivityType::Selectable )
-    MRVIEWER_API static const std::vector<std::shared_ptr<Object>>& getAllObjects();
-    // get all selectable object depth
+    // same as getAllObjectsInTree<ObjectType>( &SceneRoot::get(), SelectivityType ) but use cached data
+    template <typename ObjectType, ObjectSelectivityType SelectivityType>
+    static const std::vector<std::shared_ptr<ObjectType>>& getAllObjects();
+
+    // get all selectable object depths
     // metadata for drawing scene objects list
     MRVIEWER_API static const std::vector<int>& getAllObjectsDepth();
-    // get all selected object in tree
-    // same as getAllObjectsInTree<Object>( &SceneRoot::get(), ObjectSelectivityType::Selected )
-    MRVIEWER_API static const std::vector<std::shared_ptr<Object>>& getSelectedObjects();
 private:
-    static SceneCache& instance_();
+    MRVIEWER_API static SceneCache& instance_();
     SceneCache() {};
 
-    static void updateAllObjectsWithDepth_();
+    using StoredType = std::vector<std::shared_ptr<Object>>;
+    using CachedStoredType = std::shared_ptr<StoredType>;
+    MRVIEWER_API static std::pair<StoredType, std::vector<int>> updateAllObjectsWithDepth_();
+
+    std::vector<CachedStoredType> cachedData_;
+    std::shared_ptr<std::vector<int>> allObjectDepths_;
+
+    // Helper class to convert template params to unique numbers
+    class TypeMap
+    {
+    public:
+        template <typename ObjectType, ObjectSelectivityType SelectivityType>
+        static int getId()
+        {
+            static int myId = instance_().idCounter_++;
+            return myId;
+        }
```

Comment:
typeid ?

Reply:
it is possible, i think current version is more simple .(no need use hashmap instead of vector to storage data)